Update installation instructions to use GHCup (#1101)
exercism/haskell: docs/INSTALLATION.md

# Installation

## Get GHCup

The recommended way of installing Haskell tools is through [GHCup][ghcup].

With GHCup you can manage your installed compilers, project/package managers, and language server.

## Install Stack

For doing Exercism exercises you only need Stack.

To install Stack, simply run the GHCup installation command or script and follow the defaults but specify that you would like to have Stack installed as well.

> Did you miss the opportunity? No worries: just run `ghcup install stack` afterwards.

Reviewer: I would prefer one sentence per line. With one sentence per line diffs will be more appropriately scoped when only one sentence is changed.

GHCup will install

- **A compiler** (GHC: the Glasgow Haskell Compiler)
  
  Strictly speaking you do not need this for Exercism, but it comes in handy when you want to play with GHCi outside of exercise directories.

- **Cabal**

  A project/package manager. You do not need this for Exercism.

- **Stack**

  A project/package manager that you do need for Exercism. It will make sure you have the right compiler and libraries for each individual project (exercise).
  

> To check that installation was successful, run the command `ghcup list` and/or `stack --version`.

Reviewer: Is it self-evident what the student should be looking for as a result of the `ghcup list` and/or `stack --version` commands? If it is not, we will want to explain it here

Author: Throughout all of the instructions, we already implicitly assume that the student is familiar with the command line. So for `stack --version` I think we're good. As for `ghcup list`&hellip; see screenshot below. I do not know what failure might look like (other than the usual _command not found_).

<details><summary>Screenshot of <code>ghcup list</code></summary>

![image](https://user-images.githubusercontent.com/19817960/187233825-d9c70fd5-ae71-4291-9c7e-9e7f8384c0d1.png)
</details>

If you still think we need to be more explicit, I'll add explanations.

Reviewer: It appears I need remember better what sort of background knowledge we assume for those who are joining a track. As you can see, that greatly affects what sort of detail we need to provide. In light of the assumption you've now informed me of, I now see that we can leave it as is.

## Get the right GHC version

Stack will automatically download and install GHC when it is required. You can start the installation and test
by running the following command in a terminal:

```bash
# move into exercise directory
cd /exercism/haskell/hello-world
# start interactive environment
stack ghc -- --version
```

If this outputs something like

```
The Glorious Glasgow Haskell Compilation System, version 8.8.4
```

then GHC is installed correctly.

## Set up your editor to work with Haskell

> Haskell support is provided by the Haskell Language Server (HLS).

Reviewer: I'm just a dinosaur who never learned about this language server. But, I would not want to imply that you need the language server to have an editor that can work with Haskell.

> Haskell support is provided by the Haskell Language Server (HLS).

Reviewer: I'm just a dinosaur who never learned about this language server. But, I would not want to imply that you need the language server to have an editor that can work with Haskell, which "support is provided by..." seems to imply. It is okay if the docs *recommend* using the language server if it materially improves student experience, but I want it to be clear that it is a recommendation, not a requirement.

### Visual Studio Code

> Just get [this extension][vscode-haskell]. It will use GHCup to install HLS automatically, after which you'll be all set.

Reviewer: ```suggestion
Just get the [Haskell for Visual Studio Code extension][vscode-haskell]. It will use GHCup to install HLS automatically, after which you'll be all set.
```

By my reading, a link like "this extension" is like "click here", which I was taught by https://webaccess.berkeley.edu/ask-pecan/click-here to avoid, so I have decided to suggest this.

**Note:** with freshly downloaded exercises, HLS may have trouble starting up. If this happens, just run `stack test` and restart the language server.

### Other editors

> See the [HLS documentation on editor configuration][hls-editor-config].

Reviewer: I'm conflicted. On the one hand, the student may get confused about "what is Haskell Language Server and do I need this?". See how a lot of the install instructions also include instructions to install a language server protocol implementation, which is not in line with the stated purpose of this link (other editors). On the other hand, I wanted to suggest https://wiki.haskell.org/Editors, but that provides no commentary about each editor. Thus I don't think either of these suggestions is perfect. I would like to figure out which will be most helpful to the students.

# Installation

## Get GHCup

> The recommended way of installing Haskell tools is through [GHCup][ghcup].

Reviewer 2: We might also reference where this information comes from, the recommendation stated at https://www.haskell.org/downloads/

This can mean that when updates and changes happen, we are less likely to be outdated.

With GHCup you can manage your installed compilers, project/package managers, and language server.

## Install Stack

For doing Exercism exercises you only need Stack.

> To install Stack, simply run the GHCup installation command or script and follow the defaults but specify that you would like to have Stack installed as well.

Reviewer 2: ```suggestion
To install Stack, run the GHCup installation command or script and follow the defaults but specify that you would like to have Stack installed as well.
```

This is a complex sentence and "simply" is not implied, as we have to make decisions here.  Often "simply" is insulting to see when and if we have a problem.  We can avoid the "insult" by removing this one word.  Otherwise it does encourage one to believe that they have not understood the "simple" and are somehow the problem.

> Did you miss the opportunity?

Reviewer 2: The opportunity for what?

Reviewer 2: Still pending…

No worries: just run `ghcup install stack` afterwards.

GHCup will install

- **A compiler** (GHC: the Glasgow Haskell Compiler)

>   Strictly speaking you do not need this for Exercism, but it comes in handy when you want to play with GHCi outside of exercise directories.

Reviewer 2: Then, we probably should not mention it here for installation for Exercism.

Maybe point to resources about "learning more" elsewhere?

- **Cabal**

  A project/package manager.

>   You do not need this for Exercism.

Reviewer 2: Then do we need to mention this section?

Reviewer 2: Should we move the "not needed" things to a "Recommended" section if it makes things nicer, but is not necessary?

- **Stack**

>   A project/package manager that you do need for Exercism.

Reviewer 2: By being here, implies that we do need it, I think, if we do not mention things that we do not need, in this document.  I am not sure that this instills confidence in other places if we mention it is needed, and do not mention explicitly the need in other sections.

It is easier to state what is needed, and to state and then declare what is not needed.

Or perhaps put optional information in a complete section for those things that are nice but not necessary?

Did you miss the opportunity?
No worries: just run `ghcup install stack` afterwards.

GHCup will install

- **A compiler** (GHC: the Glasgow Haskell Compiler)

  Strictly speaking you do not need this for Exercism, but it comes in handy when you want to play with GHCi outside of exercise directories.

- **Cabal**

  A project/package manager.
  You do not need this for Exercism.

- **Stack**

If this outputs something like `ghc: no input files`, GHC is installed
correctly.
  A project/package manager that you do need for Exercism.
  It will make sure you have the right compiler and libraries for each individual project (exercise).

Note that there are other ways to install GHC on a system (like the
[Haskell platform](https://www.haskell.org/platform/)), but for
running the Exercism tests, a regular stack installation is most
likely to work correctly.
To check that installation was successful, run the command `ghcup list` and/or `stack --version`.

## Installing HLint (optional)

HLint is a tool for suggesting possible improvements to Haskell code. These
suggestions include ideas such as using alternative functions, simplifying

> code and spotting redundancies.

Reviewer: Discuss about the removal of HLint install instructions. Is it implicitly included with the editor support/Haskell Language Server? Is it deemed no longer helpful to students?

### Visual Studio Code

Just get the [Haskell extension for Visual Studio Code][vscode-haskell].
It will use GHCup to install HLS automatically, after which you'll be all set.

**Note:** with freshly downloaded exercises, HLS may have trouble starting up.
If this happens, just run `stack test` and restart the language server.

### Other editors

See the [HLS documentation on editor configuration][hls-editor-config].

> You might need to install HSL manually.

Reviewer: ```suggestion
You might need to install HLS manually.
```

Of course, HSL would be hue-sat-lum, the notation for specifying colours. We need HLS (Haskell Language Server) here

> To do so, simply run `ghcup install hls`.

Reviewer 2: ```suggestion
To do so, run `ghcup install hls`.
```

For the same reasons mentioned earlier.  Placing this one word makes one feel like they can not do the simple things, and can make one reluctant to ask for help, because it is already indicated that this is simple, so maybe if they can not understand the simple things, they have no business doing the more complicated.

Reviewer 2: I approved, even though I do not like the "simply run" combination.  It should not block.

Reviewer: other words to watch out for are "just", as in someone suggesting "why don't you just _____?"
